perfkitbenchmarker/benchmarks/nginx_wrk_benchmark.py

```diff
@@ -0,0 +1,140 @@
+                   'p50 / (1 connection p50) > nginx_wrk_max_p50_ratio')
+
+FLAGS = flags.FLAGS
+
+BENCHMARK_NAME = 'nginx_wrk'
+BENCHMARK_CONFIG = """
+nginx_wrk:
+  description: Run wrk against nginx.
+  vm_groups:
+    server:
+      vm_spec: *default_single_core
+    client:
+      vm_spec: *default_single_core
+"""
+
+
+def GetConfig(user_config):
+  return configs.LoadConfig(BENCHMARK_CONFIG, user_config, BENCHMARK_NAME)
+
+
+def _PrepareServer(vm):
+  """Installs nginx on the server."""
+  vm.Install('nginx')
+  vm.RemoteCommand('echo OK | sudo tee /usr/share/nginx/html/OK')
+  nginx.Start(vm)
+
+
+def _PrepareClient(vm):
+  """Install WRK on the client VM."""
+  vm.Install('wrk')
+
+
+def Prepare(benchmark_spec):
+  """Install nginx on one VM and wrk on another.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  nginx_vm = benchmark_spec.vm_groups['server'][0]
+  wrk_vm = benchmark_spec.vm_groups['client'][0]
+
+  vm_util.RunThreaded((lambda f: f()),
+                      [functools.partial(_PrepareServer, nginx_vm),
+                       functools.partial(_PrepareClient, wrk_vm)])
+
+
+def Run(benchmark_spec):
+  """Run wrk against nginx.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+
+  Returns:
+    A list of sample.Sample objects.
+  """
+  nginx_vm = benchmark_spec.vm_groups['server'][0]
+  wrk_vm = benchmark_spec.vm_groups['client'][0]
+  samples = []
+  errors = 0
+  connections = 1
+  baseline_p50 = None
+  p50_multiplier = FLAGS.nginx_wrk_max_p50_ratio
+
+  target = urlparse.urljoin('http://' + nginx_vm.ip_address, 'OK')
+
+  while True:
+    run_samples = list(wrk.Run(wrk_vm, connections=connections, target=target))
+    samples.extend(run_samples)
+
+    errors = next(i.value for i in run_samples if i.metric == 'errors')
+    p50 = next(i.value for i in run_samples
+               if i.metric == 'p50 latency')
+    if baseline_p50 is None:
+      baseline_p50 = p50
+
+    logging.info('Ran with %d connections; %d errors, %gms p50 '
+                 '(baseline: %gms)',
+                 connections, errors, p50, baseline_p50)
+
+    # Check stopping conditions.
+    if errors > 0:
+      logging.info('%d requests had errors. Terminating.', errors)
+    if p50 / baseline_p50 > p50_multiplier:
+      logging.info('p50 %fms is %fx baseline (%fms). Terminating.',
+                   p50, p50 / baseline_p50, baseline_p50)
+      break
+
+    # Retry with double the connections
+    connections *= 2
+
+  for sample in samples:
+    sample.metadata.update(ip_type='external')
+
+  return samples
+
+
+def Cleanup(benchmark_spec):
+  """Remove nginx and wrk.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  vms = benchmark_spec.vms
+  nginx.Stop(vms[0])
```

Comment:
You should get the nginx vm by its group. There are no guarantees about the ordering of groups in benchmark_spec.vms